modules/imgproc/src/sumpixels.avx512_skx.cpp

```diff
@@ -0,0 +1,259 @@
+        // the line that no longer contains a full 64 byte chunk.  Other code will handle the last part.
+        const int num_chunks = num_ops_in_line / 64;
+        for (int index_64byte_chunk = 0; index_64byte_chunk < num_chunks; index_64byte_chunk++){
+            integral_64_operations_avx512((__m512i *) srcs,
+                                          (__m512i *) sums,  (__m512i *) sums_above,
+                                          (__m512i *) sqsum, (__m512i *) sqsum_above,
+                                          0xFFFFFFFFFFFFFFFF, sum_accumulator, sqsum_accumulator);
+            srcs+=64; sums+=64; sums_above+=64;
+            if (sqsum){ sqsum+= 64; sqsum_above+=64; }
+        }
+    }
+
+
+    CV_ALWAYS_INLINE
+    void post_aligned_integral(const uchar *srcs,
+                               const double *sums,   const double *sums_above,
+                               const double *sqsum,  const double *sqsum_above,
+                               __m512i &sum_accumulator, __m512i &sqsum_accumulator,
+                               int num_ops_in_line) const
+    {
+        // This function handles the last few straggling operations that are not a full chunk of 64 operations
+        // We use the same algorithm, but we calculate a different operation mask.
+        const unsigned int num_operations = (unsigned int) num_ops_in_line - (num_ops_in_line / 64 * 64);
+        __mmask64 operation_mask = (1ULL<<num_operations) - 1ULL;
+
+        integral_64_operations_avx512((__m512i *) srcs, (__m512i *) sums, (__m512i *) sums_above,
+                                      (__m512i *) sqsum, (__m512i *) sqsum_above,
+                                      operation_mask, sum_accumulator, sqsum_accumulator);
+    }
+
+
+
+    CV_ALWAYS_INLINE
+    void integral_64_operations_avx512(const __m512i *srcs,
+                                       __m512i *sums,       const __m512i *sums_above,
+                                       __m512i *sqsums,     const __m512i *sqsums_above,
+                                       __mmask64 data_mask,
+                                       __m512i &sum_accumulator, __m512i &sqsum_accumulator) const
+    {
+        __m512i src_64byte_chunk = read_64_bytes(srcs, data_mask);
+
+        for(int num_16byte_chunks=0; num_16byte_chunks<4; num_16byte_chunks++) {
+            __m128i src_16bytes = _mm512_extracti64x2_epi64(src_64byte_chunk, 0x0); // Get lower 16 bytes of data
+
+            for (int num_8byte_chunks = 0; num_8byte_chunks < 2; num_8byte_chunks++) {
+
+                __m512i src_longs = convert_lower_8bytes_to_longs(src_16bytes);
+
+                // Calculate integral for the sum on the 8 entries
+                integral_8_operations(src_longs, sums_above, data_mask, sums, sum_accumulator);
+                sums++; sums_above++;
+
+                if (sqsums){ // Calculate integral for the sum on the 8 entries
```

Comment:
I think it would be better to move condition check outside of the line processing loop

Reply:
I agree in principle, but this will require some significant restructuring for what I suspect is little gain.  

Since that the condition is not data driven (the value never changes) and that the comparison is versus 0, it is 100% predictable by any modern branch predictor and so the perf implication is only the extra rob and front-end queue entries for the branch and integer compare.  Even the integer compare is not significant here since the major work being done by the algorithm is on the SSE side so it does not compete for resources for the 1 cycle subtract.

Before my next release I'll do a rewrite of this to move the compare outside the main loop and split the algorithm to measure exactly what the performance implications are for this extra branch and report back as to whether the performance difference is worth the added complexity.  

Is this an acceptable resolution?

Reply:
I agree that result of this change could be quite small, so it could be postponed.

Reply:
Hey Vitaly,
   Just to close the loop on the discussion about moving the sqsum conditional check outside of the inner loop.  I moved it all the way out to the main caller and created a different object based on sqsum (at opt_AVX512_SKX::calculate_integral_avx512 at the bottom of the file).  So the algorithm was duplicated for sum and sqsum with sum not having any dependency on sqsum.

Then I compared this versus the current implementation with the branch in place where it is now.  Here is the results

|Name of Test|With Inner Branch|Without Inner Branch|With vs without branch(x-factor)|
|---|:-:|:-:|:-:|
|integral(127x61, 8UC3, CV_64F)|0.009|0.009|0.99|
|integral(640x480, 8UC3, CV_64F)|0.376|0.358|1.05|
|integral(1280x720, 8UC3, CV_64F)|1.497|1.441|1.04|
|integral(1920x1080, 8UC3, CV_64F)|4.087|3.957|1.03|
|integral_sqsum(127x61, 8UC3, CV_64F)|0.017|0.017|1.00|
|integral_sqsum:(640x480, 8UC3, CV_64F)|0.955|0.926|1.03|
|integral_sqsum(1280x720, 8UC3, CV_64F)|4.100|4.009|1.02|
|integral_sqsum(1920x1080, 8UC3, CV_64F)|9.556|9.415|1.01|

These values seem to fluctuate a little from run-to-run, but basically shows a tiny improvement.  Obviously the branch predictor was very good at dealing with this.  If I can find a way to implement this without causing the code to become difficult to maintain I will.